400-claims.md

```diff
@@ -128,64 +128,105 @@ The parameter data stored in a claim data is _the resolved key/value pairs_ that
   - Valid user-supplied values are presented
   - Default values are supplied for all parameters where `default` is provided and no user-supplied value overrides this
 
-### Outputs
+### Claim Results
+
+The result of executing an operation is defined separately in a claim result document. A claim may have multiple results. For example if the operation was cancelled and then retried the claim will have two results: one representing the cancelled operation and another the retried operation.
 
-The claim `result` provides a list of output names generated by the operation. A tool may choose to request the contents of any outputs to persist them but is not required to do so.
+```json
+{
+  "claimRef": {
+    "installation": "technosophos.helloworld",
+    "revision": "01CP6XM0KVB9V1BQDZ9NK8VP29"
+  },
+  "custom": {},
+  "id": "01E2ZZ2FKSE0V41DCXFCSW5D1M",
+  "created": "2018-08-30T20:39:55.549002887-06:00",
+  "modified": "2018-08-30T20:39:59.611068556-06:00",
+  "message": "",
+  "outputs": {
+    "clientCert": {},
+    "hostName": {},
+    "port": {}
+  },
+  "started": "2018-08-30T20:39:55.549002887-06:00",
+  "status": "succeeded",
+  "stopped": "2018-08-30T20:39:59.611068556-06:00"
+}
+```
+Source: [400.01-claim-result.json](examples/400.01-claim-result.json)
+
+The fields above are defined as follows:
+
+- `claimRef` (REQUIRED): Reference to the claim that generated this result.
+  - `installation` (REQUIRED): The claim installation name.
+  - `revision` (REQUIRED): The claim revision.
+- `custom` (OPTIONAL): A section for custom extension data applicable to a given runtime.
+- `id` (REQUIRED): An [ULID](https://github.com/ulid/spec).
+- `created` (REQUIRED): A timestamp indicating when this result was first created. This MUST not be changed after initial creation.
+- `modified` (REQUIRED): A timestamp indicating the last time this result was modified. Creating the result MUST set this to a timestamp that matches the result's `created`.
+- `message` (OPTIONAL): A human-readable string that communicates the outcome. Error messages MAY be included in `failed` conditions.
+- `outputs` (OPTIONAL): Outputs generated by the operation. It is a map from the output names to metadata about the output. The  output value is not stored in the claim result. See the [Outputs](#outputs) section for details. If this field is not present, it may be assumed that no outputs were generated as a result of the operation.
+- `started` (OPTIONAL): A timestamp indicating when execution of the operation started.
```

Comment:
Should we allow results with final statuses, like `succeeded` or `failed` to set this field? Would it be valid for other "still running" statuses to have the `stopped` field set?

Reply:
Sorry I missed this comment! 😊 Are you asking about `stopped` or `started`? I had intended for result to be mutable, so as the operation proceeded the tool would usually look like this:

1. Create the result and set created, modified to the same time. If it uses queuing, set status to `pending`. 
2. When execution starts set `started` and the status to `running`. At this point stopped is null.
3. When execution completes set `stopped` and status to `failed` or `succeeded`.

Based on how the tools works, some of these may occur in a single update to the result or in separate steps, but anytime it modifies the result, `modified` is bumped.

Does that help clarify? Does any of the wording in the PR need to change to reflect what I just said in the example above?

Reply:
That makes sense. I'm still wondering if we could still satisfy everyone's use cases without created, updated, started, and stopped timestamps all on the result object.

I would prefer to have a single `observed` timestamp and treat results as a list of events associated with a claim, but even if we don't require implementations to go with that model it seems like with created and modified could always be derived from started and stopped (or vice-versa).

Reply:
Are you suggesting that all implementations would have to make a single result document per "event", e.g. the operation is pending is one result, when the execution starts that's another result, etc. So the result document wouldn't be mutable either?

Reply:
Yes, this would also prevent runtimes from overwriting potentially important metadata for debugging purposes, like exactly when an output file was generated or the `result.message` for some intermediate step in the claim's lifecycle.

Runtimes could choose to persist only the first and last observed results, which would end up being pretty much the same as using a single mutable result and continuously updating its fields.

Reply:
Gotcha! 👍 That seems fine with me. This pushes the claim spec much closer to laying out an implementation but honestly I think that is a good thing.